### ISSUE TITLE
remove_iiv

### ISSUE THREAD

**Issue by SCarter79** [pharmpy/pharmpy#117]
Pharmpy 0.17.0, running python 3.8.7 with R 4.0.3 in R studio 1.3.1056. If i remove an ETA, the number of ETAs do not automatically update to exclude that ETA, model code from R:

```
model30 <- model20$copy() %>%
  modeling$unfix_parameters(c('POP_CL',
                              'POP_VC',
                              'POP_MAT',
                              'POP_QP1',
                              'POP_VP1',
                              'POP_MDT')) %>%
  modeling$remove_iiv(list('ETA(4)')) %>%
  modeling$update_source() %>%
  modeling$write_model(here::here('run30.mod'), force = TRUE)

```

NONMEM output:

```
$SUBROUTINES ADVAN5  TRANS1 
$MODEL COMPARTMENT=(TRANSIT1 DEFDOSE) COMPARTMENT=(TRANSIT2) COMPARTMENT=(TRANSIT3) COMPARTMENT=(DEPOT) COMPARTMENT=(CENTRAL) COMPARTMENT=(PERIPHERAL)
$PK
MDT = THETA(6)*EXP(ETA(6))
VP1 = THETA(5)*EXP(ETA(5))
QP1 = THETA(4)
CL = THETA(1) * EXP(ETA(1))
VC = THETA(2) * EXP(ETA(2))
MAT = THETA(3) * EXP(ETA(3))

```

VP1 should update to ETA(4): `VP1 = THETA(5)*EXP(ETA(4))`
MDT should update to ETA(5): `MDT = THETA(6)*EXP(ETA(5))`

run20.mod:

```
$PROBLEM
$INPUT ID VISI DAT1 DGRP DOSE FLAG ONO XIME DVO NEUY SCR AGE SEX NYHA WT COMP IACE DIG DIU NUMB TAD TIME VID CRCL AMT SS II VID1 CMT CONO DV EVID OVID SHR SHR2
$DATA mx19B.csv IGNORE=@
$SUBROUTINES ADVAN5  TRANS1 
$MODEL COMPARTMENT=(TRANSIT1 DEFDOSE) COMPARTMENT=(TRANSIT2) COMPARTMENT=(TRANSIT3) COMPARTMENT=(DEPOT) COMPARTMENT=(CENTRAL) COMPARTMENT=(PERIPHERAL)
$PK
MDT = THETA(6)*EXP(ETA(6))
VP1 = THETA(5)*EXP(ETA(5))
QP1 = THETA(4)*EXP(ETA(4))
CL = THETA(1) * EXP(ETA(1))
VC = THETA(2) * EXP(ETA(2))
MAT = THETA(3) * EXP(ETA(3))
KA = 1/MAT
V2 = VC
Q = QP1
V3 = VP1
K12 = 3/MDT
K23 = 3/MDT
K34 = 3/MDT
K45 = KA
K56 = Q/V2
K50 = CL/V2
K65 = Q/V3
$ERROR
CONC = A(5)/VC
Y = CONC + CONC * EPS(1)
$ESTIMATION METHOD=1 INTER
$COVARIANCE PRINT=E
$THETA (0, 22.4506, Inf) FIX ; POP_CL
$THETA (0, 107.372, Inf) FIX ; POP_VC
$THETA (0, 0.311228, Inf) FIX ; POP_MAT
$THETA  (0,10.346) FIX ; POP_QP1
$THETA  (0,40.6582) FIX ; POP_VP1
$THETA  (0,0.125) FIX ; POP_MDT
$OMEGA 0.001; IIV_CL
$OMEGA 0.001; IIV_VC
$OMEGA 0.001; IIV_MAT
$OMEGA  0.001 ; IIV_QP1
$OMEGA  0.001 ; IIV_VP1
$OMEGA  0.001 ; IIV_MDT
$SIGMA 0.264233; RUV_PROP

```

**Comment by rikardn** [comment]
I cannot reproduce this issue with Pharmpy 0.18.0. @stellabelin has this been fixed? I test the python code below (the mx19B was taken from @SCarter79) and get the correct numbering.

```python
from pharmpy import Model
from io import StringIO

code = """$PROBLEM
$INPUT ID VISI DAT1 DGRP DOSE FLAG ONO XIME DVO NEUY SCR AGE SEX NYHA WT COMP IACE DIG DIU NUMB TAD TIME VID CRCL AMT SS II VID1 CMT CONO DV EVID OVID SHR SHR2
$DATA /path/to/mx19B.csv IGNORE=@
$SUBROUTINES ADVAN5  TRANS1
$MODEL COMPARTMENT=(TRANSIT1 DEFDOSE) COMPARTMENT=(TRANSIT2) COMPARTMENT=(TRANSIT3) COMPARTMENT=(DEPOT) COMPARTMENT=(CENTRAL) COMPARTMENT=(PERIPHERAL)
$PK
MDT = THETA(6)*EXP(ETA(6))
VP1 = THETA(5)*EXP(ETA(5))
QP1 = THETA(4)*EXP(ETA(4))
CL = THETA(1) * EXP(ETA(1))
VC = THETA(2) * EXP(ETA(2))
MAT = THETA(3) * EXP(ETA(3))
KA = 1/MAT
V2 = VC
Q = QP1
V3 = VP1
K12 = 3/MDT
K23 = 3/MDT
K34 = 3/MDT
K45 = KA
K56 = Q/V2
K50 = CL/V2
K65 = Q/V3
$ERROR
CONC = A(5)/VC
Y = CONC + CONC * EPS(1)
$ESTIMATION METHOD=1 INTER
$COVARIANCE PRINT=E
$THETA (0, 22.4506, Inf) FIX ; POP_CL
$THETA (0, 107.372, Inf) FIX ; POP_VC
$THETA (0, 0.311228, Inf) FIX ; POP_MAT
$THETA  (0,10.346) FIX ; POP_QP1
$THETA  (0,40.6582) FIX ; POP_VP1
$THETA  (0,0.125) FIX ; POP_MDT
$OMEGA 0.001; IIV_CL
$OMEGA 0.001; IIV_VC
$OMEGA 0.001; IIV_MAT
$OMEGA  0.001 ; IIV_QP1
$OMEGA  0.001 ; IIV_VP1
$OMEGA  0.001 ; IIV_MDT
$SIGMA 0.264233; RUV_PROP"""

model = Model(StringIO(code))

from pharmpy.modeling import *

unfix_parameters(model, ['POP_CL', 'POP_VC', 'POP_MAT', 'POP_QP1', 'POP_VP1', 'POP_MDT'])
remove_iiv(model, ['ETA(4)'])
model.update_source()
print(model)
```

**Comment by stellabelin** [comment]
I can reproduce it in 0.17.0, I think it has something to do with the fact that update_source is called twice (via update_source and then write_model). In 0.19.0 it seems to work (maybe due to the fact that I reworked parts of update_random_variables in 77ac9e75ab7678f904363dbfc75b43f3015938a5) using this:

```
unfix_parameters(model, ['POP_CL', 'POP_VC', 'POP_MAT', 'POP_QP1', 'POP_VP1', 'POP_MDT'])
remove_iiv(model, 'ETA(4)')
model.update_source()
model.update_source()
print(model)
```

**Comment by SCarter79** [comment]
It only calls the update source once i think in the code i sent? i'll see if it still does it in 0.18.0

**Comment by stellabelin** [comment]
It calls update_source in write_model as well! So yes, your code itself only called it once but when you call write_model it in turn will call update_source.

**Comment by rikardn** [comment]
Ok, so not in the latest version. Close?

**Comment by SCarter79** [comment]
all fixed in 0.19.0 thanks!
